core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -672,6 +672,22 @@ private[spark] class ExternalSorter[K, V, C](
     partitionedIterator.flatMap(pair => pair._2)
   }
 
+  /**
+   * Return an iterator over all the data written to this object, aggregated by our aggregator.
+   * On task completion (success, failure, or cancellation), it updates related task metrics,
+   * and releases resources by calling `stop()`.
+   */
+  def completionIterator: Iterator[Product2[K, C]] = {
+    context.taskMetrics().incMemoryBytesSpilled(memoryBytesSpilled)
+    context.taskMetrics().incDiskBytesSpilled(diskBytesSpilled)
+    context.taskMetrics().incPeakExecutionMemory(peakMemoryUsedBytes)
```

Comment:
I think metrics updates implicitly require records inserting, otherwise, it's meaningless to update. So, shall we do `insertAll` here too? e.g.,

```scala
def completionIterator(records: Iterator[Product2[K, V]]): Iterator[Product2[K, C]] = {
  insertAll(records)
  context.taskMetrics().incMemoryBytesSpilled(memoryBytesSpilled)
  ...
}
```

Reply:
I am ok to add `records` here, but perfer another method name 'insertAllAndUpdateMetrics'

Reply:
Yeah, I was thinking about the renaming too. Maybe, `insertAllWithMetricsUpdated`?

Reply:
Ah, I just saw you already renamed the method name. Should be good enough. You can ignore mine.